app/product_notices.go

```diff
@@ -0,0 +1,260 @@
+		if !validateConfigEntry(a, k, v) {
+			return false, nil
+		}
+	}
+
+	// check the type of installation
+	if cnd.InstanceType != nil {
+		if !cnd.InstanceType.Matches(isCloud) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func validateConfigEntry(a *App, path string, expectedValue interface{}) bool {
+	value, found := config.GetConfigValueByPath(a.Config(), strings.Split(path, "."))
+	if !found {
+		return false
+	}
+	vt := reflect.ValueOf(value)
+	if vt.IsNil() {
+		return expectedValue == nil
+	}
+	if vt.Kind() == reflect.Ptr {
+		vt = vt.Elem()
+	}
+	val := vt.Interface()
+	return val == expectedValue
+}
+
+func (a *App) GetProductNotices(lastViewed int64, userId, teamId string, client model.NoticeClientType, clientVersion string, locale string) (model.NoticeMessages, *model.AppError) {
+	isSystemAdmin := a.SessionHasPermissionTo(*a.Session(), model.PERMISSION_MANAGE_SYSTEM)
+	isTeamAdmin := a.SessionHasPermissionToTeam(*a.Session(), teamId, model.PERMISSION_MANAGE_TEAM)
+
+	// check if notices for regular users are disabled
+	if !*a.Srv().Config().AnnouncementSettings.UserNoticesEnabled && !isTeamAdmin && !isSystemAdmin {
+		return []model.NoticeMessage{}, nil
+	}
+
+	// check if notices for admins are disabled
+	if !*a.Srv().Config().AnnouncementSettings.AdminNoticesEnabled && (isTeamAdmin || isSystemAdmin) {
+		return []model.NoticeMessage{}, nil
+	}
+
+	views, err := a.Srv().Store.ProductNotices().GetViews(userId)
+	if err != nil {
+		return nil, model.NewAppError("GetProductNotices", "api.system.update_viewed_notices.failed", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	sku := a.Srv().ClientLicense()["SkuShortName"]
+	isCloud := a.Srv().ClientLicense()["Cloud"] != ""
+
+	var filteredNotices []model.NoticeMessage
+
+	getViewState := func(nid string) *model.ProductNoticeViewState {
+		for _, v := range views {
+			if v.NoticeId == nid {
+				return &v
+			}
+		}
+		return nil
+	}
+
+	for _, notice := range cachedNotices {
+		// check if the notice has been viewed already
+		view := getViewState(notice.ID)
+		if view != nil {
+			repeatable := notice.Repeatable != nil && *notice.Repeatable
+			if !repeatable && view.Viewed > 0 {
+				continue
+			}
+			if repeatable && view.Viewed > MAX_REPEAT_VIEWINGS {
+				continue
+			}
+			if view.Timestamp < lastViewed {
+				continue
+			}
+		}
+		notice := notice // pin
+		result, err := noticeMatchesConditions(a,
+			client,
+			clientVersion,
+			locale,
+			cachedPostCount,
+			cachedUserCount,
+			isSystemAdmin,
+			isTeamAdmin,
+			isCloud,
+			sku,
+			&notice)
+		if err != nil {
+			return nil, model.NewAppError("GetProductNotices", "api.system.update_notices.validating_failed", nil, err.Error(), http.StatusBadRequest)
+		}
+		if result {
+			selectedLocale := "enUS"
+			filteredNotices = append(filteredNotices, model.NoticeMessage{
+				NoticeMessageInternal: notice.LocalizedMessages[selectedLocale],
+				ID:                    notice.ID,
+			})
```

Comment:
Doesn't this mean the returned filtered notices will all be in `enUS` instead of the requested locale?

Reply:
@ashishbhate correct. in this phase - locale is ignored